[InstCombine] Fold ((X << nuw Z) binop nuw Y) >>u Z --> X binop nuw (Y >>u Z)

AZero13 opened the pull request:
Proofs:
https://alive2.llvm.org/ce/z/N9dRzP
https://alive2.llvm.org/ce/z/Xrpc-Y
https://alive2.llvm.org/ce/z/BagBM6
https://alive2.llvm.org/ce/z/k2dtYZ

goldsteinn commented on llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:
```diff
@@ -1267,6 +1267,19 @@ Instruction *InstCombinerImpl::visitLShr(BinaryOperator &I) {
 
+  // Special Case:
+  // if both the add and the shift are nuw, we can omit the AND entirely
+  // ((X << Y) nuw + Z nuw) >>u Z --> (X + (Y >>u Z))
```
do you mean `) nuw >> u Z`?

AtariDreams replied:
Yeah

goldsteinn replied:
Also do you mean `((X << Z) + Y) >> Z`?

AtariDreams replied:
No.

goldsteinn commented on llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:
```diff
@@ -1267,6 +1267,19 @@ Instruction *InstCombinerImpl::visitLShr(BinaryOperator &I) {
+  Value *Y;
+  if (match(Op0, m_OneUse(m_c_NUWAdd((m_NUWShl(m_Value(X), m_Specific(Op1))),
+                                     m_Value(Y))))) {
+    Value *NewLshr = Builder.CreateLShr(Y, Op1, "", I.isExact());
```
This exact is only valid if `I` is `exact`

AtariDreams replied:
I know. Thats' why I.isExact is there

goldsteinn replied:
Oh i misread. Sorry.

Can you please update the proof to include the non-exact version and non-nsw version.

AtariDreams replied:
Sure: https://alive2.llvm.org/ce/z/wORNYm

AtariDreams commented:
@nikic What do you think about this?

AtariDreams commented:
@arsenm What do you think about this PR?

arsenm commented on llvm/test/Transforms/InstCombine/lshr.ll:
```diff
@@ -360,8 +360,64 @@ define <3 x i14> @mul_splat_fold_vec(<3 x i14> %x) {
+; CHECK-NEXT:    [[TMP3:%.*]] = add nuw nsw i32 [[TMP2]], [[TMP1]]
+; CHECK-NEXT:    ret i32 [[TMP3]]
+;
+  %2 = mul nsw nuw i32 %x, 3
```
Use named values in tests

arsenm commented on llvm/test/Transforms/InstCombine/lshr.ll:
```diff
@@ -360,8 +360,64 @@ define <3 x i14> @mul_splat_fold_vec(<3 x i14> %x) {
+  ret i32 %3
+}
+
+  define i32 @shl_add_lshr (i32 %x, i32 %c, i32 %y) {
```
Indentation off

arsenm commented on llvm/test/Transforms/InstCombine/lshr.ll:
```diff
@@ -360,8 +360,64 @@ define <3 x i14> @mul_splat_fold_vec(<3 x i14> %x) {
+  ret i32 %3
+}
+
+define i32 @ashr_mul_times_3_div_2_exact (i32 %0) {
```
Extra space before ( in all of these

arsenm commented on llvm/test/Transforms/InstCombine/lshr.ll:
```diff
@@ -360,8 +360,64 @@ define <3 x i14> @mul_splat_fold_vec(<3 x i14> %x) {
+  %3 = lshr i32 %2, 1
+  ret i32 %3
+}
+
```
Missing negative tests for multiple uses and individual flag missing cases

AtariDreams commented:
@arsenm This is ready for review and merge! (I don't have commit perms)

arsenm commented on llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:
```diff
@@ -1450,9 +1460,24 @@ Instruction *InstCombinerImpl::visitLShr(BinaryOperator &I) {
         }
+
+        // Special case: lshr nuw (mul (X, 3), 1) -> add nuw nsw (X, lshr(X, 1)
+        if (ShAmtC == 1 && MulC->getZExtValue() == 3) {
```
Is there a more general form for this than these specific constants? Why is this the one special case?

arsenm replied:
Can you add an alive2 link?

AtariDreams replied:
> Is there a more general form for this than these specific constants? Why is this the one special case? 

(X+(x<<1))>>1 is canonicalized to x*3/2 so this is missed as a result

AtariDreams replied:
> Can you add an alive2 link? 

Pretty sure I did in the first comment I made

AtariDreams commented:
@arsenm I removed the constant specific stuff and moved a generalization of that to another PR.

AtariDreams commented:
@dtcxzyw would you be so kind as to benchmark this

dtcxzyw commented on llvm/test/Transforms/InstCombine/lshr.ll:
```diff
@@ -360,8 +360,33 @@ define <3 x i14> @mul_splat_fold_vec(<3 x i14> %x) {
   ret <3 x i14> %t
 }
 
+define i32 @shl_add_lshr(i32 %x, i32 %c, i32 %y) {
```
Please rename this test to `shl_add_lshr_flag_preservation`, and add the following basic test:
```
define i32 @shl_add_lshr(i32 %x, i32 %c, i32 %y) {
   %shl = shl nuw i32 %x, %c
   %add = add nuw i32 %shl, %y
   %lshr = lshr i32 %add, %c
   ret i32 %lshr
}
```

AtariDreams commented:
@dtcxzyw I addressed all of your concerns!

arsenm commented on llvm/test/Transforms/InstCombine/lshr.ll:
```diff
@@ -360,8 +360,58 @@ define <3 x i14> @mul_splat_fold_vec(<3 x i14> %x) {
+  %res = lshr exact i32 %add, %z
+  ret i32 %res
+}
+
```
Add a vector test

arsenm commented on llvm/test/Transforms/InstCombine/lshr.ll:
```diff
@@ -360,8 +360,58 @@ define <3 x i14> @mul_splat_fold_vec(<3 x i14> %x) {
+  %lshr = lshr exact i32 %add, %c
+  ret i32 %lshr
+}
+
```
Extra newline

arsenm reviewed (approved):
test nit

AtariDreams commented:
@arsenm Done!

nikic reviewed (commented):
This looks fine, but can you please clean up the alive2 proofs in the PR description? They seem to contain a whole bunch of stuff that is not related to the fold implemented here, like things with ashr and muls.

The proofs should all have the base shl+add+lshr sequence, just with different flags.

AtariDreams commented:
@nikic @arsenm I do not have commit perms. What do you think about merging this?

AZero13 committed:
[InstCombine] Pre-commit tests (NFC)

AtariDreams commented:
@nikic @arsenm @topperc I do not have commit perms. What do you think about merging this?

nikic commented on llvm/test/Transforms/InstCombine/lshr.ll:
```diff
@@ -360,8 +371,57 @@ define <3 x i14> @mul_splat_fold_vec(<3 x i14> %x) {
+  %add = add nuw i32 %y, %shl
+  %lshr = lshr i32 %add, %c
+  ret i32 %lshr
+}
```
This test does not actually test  the commuted case, because it gets canonicalized, see https://llvm.godbolt.org/z/93sGo78W8.

Read https://llvm.org/docs/InstCombineContributorGuide.html#add-commuted-tests to understand how to resolve this issue.

Also, your pre-commit tests don't show this because your second commit modifies the test inputs as well. The split is to add all tests in the first commit and only change CHECK lines in the second commit. Do not modify test inputs in the second commit.

nikic replied:
This was, once again, ***not*** resolved. Your commutative tests are still incorrect.

AtariDreams replied:
I'll write another test.

AtariDreams commented:
@nikic I addressed the issues!

AZero13 committed:
[InstCombine] Fold ((X << nuw Z) binop nuw Y) >>u Z --> X binop nuw (Y >>u Z)

Proofs:
https://alive2.llvm.org/ce/z/N9dRzP
https://alive2.llvm.org/ce/z/Xrpc-Y
https://alive2.llvm.org/ce/z/k2dtYZ

AtariDreams commented:
@dtcxzyw can you plz run the benchmark again I wanna see something

arsenm commented on llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:
```diff
@@ -1259,6 +1259,52 @@ Instruction *InstCombinerImpl::visitLShr(BinaryOperator &I) {
+  if (match(Op0, m_OneUse(m_c_BinOp(m_NUWShl(m_Value(X), m_Specific(Op1)),
+                                    m_Value(Y))))) {
+    BinaryOperator *Op0OB = cast<BinaryOperator>(Op0);
+    bool canFold = false;
+    if (isSuitableBinOpcode(Op0OB->getOpcode())) {
+      if (auto *Inst = dyn_cast<OverflowingBinaryOperator>(Op0)) {
+        if (Inst->hasNoUnsignedWrap())
+          canFold = true;
+      } else
+        canFold = true;
+
+      if (canFold) {
```
You can avoid the canFold variable: 

if (auto *OBO = dyn_cast<OverflowingBinaryOperator>(Op0); 
    !OBO || OBO->hasNoUnsignedWrap()) {

arsenm commented on llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:
```diff
@@ -1259,6 +1259,50 @@ Instruction *InstCombinerImpl::visitLShr(BinaryOperator &I) {
     return new ZExtInst(Builder.CreateIsNotNeg(X, "isnotneg"), Ty);
 
+  Value *Y;
+  if (match(Op0, m_OneUse(m_NUWSub(m_NUWShl(m_Value(X), m_Specific(Op1)),
```
This part didn't get a pattern comment

AtariDreams replied:
Fixed! @arsenm

AtariDreams commented:
I just realized I forgot to fix the order of the sub because the order matters, we do not canonicalize it.

I fixed the miscompile, I apologize @dtcxzyw

AtariDreams commented:
Okay it is resolved now, and I added new tests @dtcxzyw, I apologize for the inconvenience.

AtariDreams commented:
@dtcxzyw Though, it was thanks to the opt-benchmark that I reviewed by changes because something was off about the diff, and it did not seem equivalent at all.

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:
```diff
@@ -1259,6 +1259,53 @@ Instruction *InstCombinerImpl::visitLShr(BinaryOperator &I) {
     return new ZExtInst(Builder.CreateIsNotNeg(X, "isnotneg"), Ty);
 
+  // ((X << nuw Z) sub nuw Y) >>u exact Z --> X sub nuw (Y >>u exact Z),
+  // ONLY if I is exact, and both the shift and sub are nuw
```
```suggestion
```
It is the same as the above line.

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:
```diff
@@ -1259,6 +1259,53 @@ Instruction *InstCombinerImpl::visitLShr(BinaryOperator &I) {
+  if (I.isExact() &&
+      match(Op0, m_OneUse(m_NUWSub(m_NUWShl(m_Value(X), m_Specific(Op1)),
+                                   m_Value(Y))))) {
+    Value *NewLshr = Builder.CreateLShr(Y, Op1, "", /* isExact */ true);
```
```suggestion
    Value *NewLshr = Builder.CreateLShr(Y, Op1, "", /*isExact=*/true);
```

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:
```diff
@@ -1259,6 +1259,53 @@ Instruction *InstCombinerImpl::visitLShr(BinaryOperator &I) {
+        if (OBO) {
+          NewBinOp->setHasNoUnsignedWrap(true);
+          NewBinOp->setHasNoSignedWrap(OBO->hasNoSignedWrap());
+        } else if (auto *Disjoint = dyn_cast<PossiblyDisjointInst>(Op0);
+                   Disjoint && Disjoint->isDisjoint()) {
+          cast<PossiblyDisjointInst>(NewBinOp)->setIsDisjoint(true);
```
```suggestion
        } else if (auto *Disjoint = dyn_cast<PossiblyDisjointInst>(Op0)) {
          cast<PossiblyDisjointInst>(NewBinOp)->setIsDisjoint(Disjoint->isDisjoint());
```

dtcxzyw commented:
@AtariDreams It doesn't modify anything :( https://github.com/llvm/llvm-project/compare/c8214019d1da25bb8211107e36461f9c1e7c764a..cdabaae43b855a47611a3a02ab2882f1dcd222dd

Don't rebase your patch if there is no conflict to be resolved.
See https://llvm.org/docs/GitHub.html#rebasing-pull-requests-and-force-pushes

goldsteinn commented on llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:
```diff
@@ -1259,6 +1259,53 @@ Instruction *InstCombinerImpl::visitLShr(BinaryOperator &I) {
+        return NewBinOp;
+      }
+    }
+  }
```
Which proofs are for this case? Seems like all your proofs have the `exact` flag.
Also missing a proof for `disjoint` propagation.

AtariDreams replied:
Am I?

dtcxzyw commented on llvm/test/Transforms/InstCombine/lshr.ll:
```diff
@@ -360,8 +371,187 @@ define <3 x i14> @mul_splat_fold_vec(<3 x i14> %x) {
   ret <3 x i14> %t
 }
 
+define i32 @shl_add_lshr_flag_preservation(i32 %x, i32 %c, i32 %y) {
```
Missing test for `add` without nuw flag.

AtariDreams replied:
Added! @dtcxzyw

dtcxzyw reviewed (approved):
LGTM.

Please don't force-push in this PR. The notifications flood my mailbox :(

goldsteinn commented on llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:
```diff
@@ -1259,6 +1259,53 @@ Instruction *InstCombinerImpl::visitLShr(BinaryOperator &I) {
+  };
+
+  // If both the binop and the shift are nuw, then:
+  // ((X << nuw Z) binop nuw Y) >>u Z --> X binop nuw (Y >>u Z)
```
Although you can handle `And`, you just cant propagate `exact`: https://alive2.llvm.org/ce/z/JNaAvP

AtariDreams commented:
@arsenm I do not have commit perms by the way.

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:
```diff
@@ -1259,6 +1259,54 @@ Instruction *InstCombinerImpl::visitLShr(BinaryOperator &I) {
+    default:
+      return false;
+    case Instruction::Add:
+    case Instruction::And:
```
Oops, this line doesn't match my reviewed version. https://github.com/llvm/llvm-project/commit/5d61e1f6285846d4599cd77dc6a873aa13901aaf

cc @nikic @arsenm @tstellar Is there a way to revoke the approval after new changes are made to this PR?

@AtariDreams It is dangerous to force-push non-NFC changes after receiving an approval.

nikic replied:
> cc @nikic @arsenm @tstellar Is there a way to revoke the approval after new changes are made to this PR?

You mean in general? Github has an option to "Dismiss stale pull request approvals when new commits are pushed", but I think this is bound to multiple layers of features we do not use (required pull requests with required approvals). It also doesn't work well with the common practice where the approval contains some final nits (usually NFC nits).

For a specific PR, you can also dismiss your old approval manually, but of course this only works if you notice the change.

AtariDreams replied:
> Oops, this line doesn't match my reviewed version. https://github.com/llvm/llvm-project/commit/5d61e1f6285846d4599cd77dc6a873aa13901aaf
> 
> 
> 
> cc @nikic @arsenm @tstellar Is there a way to revoke the approval after new changes are made to this PR?
> 
> 
> 
> @AtariDreams It is dangerous to force-push non-NFC changes after receiving an approval.
> 
> 

And was added when another maintainer told me it worked, but only if I dropped the exact flags, and I added that proof to the proofs
